### PR TITLE
Add MeasurementGate.with_key to change measurement key

### PR DESCRIPTION
**Pull request by maffoo**
The "rekey" operation is useful for some circuit transformations we want to implement; defining it on the MeasurementGate ensures that all the gate properties get copied properly when rekeying.

### CODE REVIEW

File: cirq/google/api/v2/results.py [hunk 1]
```diff
@@ -77,12 +77,11 @@ def find_measurements(program: 'cirq.Circuit',) -> List[MeasureInfo]:
 def _circuit_measurements(circuit: 'cirq.Circuit') -> Iterator[MeasureInfo]:
     for i, moment in enumerate(circuit):
         for op in moment:
-            if (isinstance(op, ops.GateOperation) and
-                    isinstance(op.gate, ops.MeasurementGate)):
+            if isinstance(op.gate, ops.MeasurementGate):
                 yield MeasureInfo(key=op.gate.key,
                                   qubits=_grid_qubits(op),
                                   slot=i,
-                                  invert_mask=_full_mask(op))
```
Comment (balopat) [inline cirq/google/api/v2/results.py:85]: This leaves `_full_mask` unused, should we remove it?

Comment (maffoo) [reply cirq/google/api/v2/results.py:85]: Good call. Done.

File: cirq/ops/measurement_gate.py [hunk 2]
```diff
@@ -81,7 +88,8 @@ def with_bits_flipped(self, *bit_positions: int) -> 'MeasurementGate':
             new_mask[b] = not new_mask[b]
         return MeasurementGate(self.num_qubits(),
                                key=self.key,
-                               invert_mask=tuple(new_mask))
+                               invert_mask=tuple(new_mask),
+                               qid_shape=self._qid_shape)
```
Comment (balopat) [inline cirq/ops/measurement_gate.py:92]: This is fine here - but it popped a question in my head which might result in a new issue: how do we (do we?) apply an "invert_mask" on qudits?

Comment (maffoo) [reply cirq/ops/measurement_gate.py:92]: I don't think we do that anywhere. Frankly I'm not sure that invert mask as a thing is particularly useful, but we added it early on (well before we had any notion of qudits in cirq).

Comment (balopat) [reply cirq/ops/measurement_gate.py:92]: Got it! 

Re qudits: I found https://github.com/quantumlib/Cirq/pull/2079 explaining that it flips |0> with |1> and leaves |2+> alone. 
Re general usefulness: It seems (at least based on https://github.com/quantumlib/Cirq/issues/1943) that some users use invert mask

**Comment by maffoo** [comment]
Is there a way to tell the cla bot to report its status?

**Pull request merged**